lib/IRGen/GenHeap.cpp

```diff
@@ -39,13 +39,226 @@
+      return IGF.get##Name##ExtraInhabitantIndex(src, \
+                                                ReferenceCounting::Nativeness); \
+    } \
+    void storeExtraInhabitant(IRGenFunction &IGF, llvm::Value *index, \
+                              Address dest, SILType T) const override { \
+      return IGF.store##Name##ExtraInhabitant(index, dest, \
+                                             ReferenceCounting::Nativeness); \
+    } \
+    APInt getFixedExtraInhabitantMask(IRGenModule &IGM) const override { \
+      return IGM.get##Name##ExtraInhabitantMask(ReferenceCounting::Nativeness); \
+    } \
+    llvm::Type *getOptionalIntType() const { \
+      return llvm::IntegerType::get( \
+          ValueTypeAndIsOptional.getPointer()->getContext(), \
+          getFixedSize().getValueInBits()); \
+    } \
+  };
+#define NEVER_LOADABLE_CHECKED_REF_STORAGE(Name, ...) \
+  NEVER_LOADABLE_CHECKED_REF_STORAGE_HELPER(Name, Native) \
+  NEVER_LOADABLE_CHECKED_REF_STORAGE_HELPER(Name, Unknown)
+#define SOMETIMES_LOADABLE_CHECKED_REF_STORAGE(Name, ...) \
+  NEVER_LOADABLE_CHECKED_REF_STORAGE_HELPER(Name, Unknown) \
+  class Native##Name##ReferenceTypeInfo \
+    : public SingleScalarTypeInfo<Native##Name##ReferenceTypeInfo, \
+                                  LoadableTypeInfo> { \
+    llvm::PointerIntPair<llvm::Type*, 1, bool> ValueTypeAndIsOptional; \
+  public: \
+    Native##Name##ReferenceTypeInfo(llvm::Type *valueType, \
+                                              llvm::Type *type, \
+                                              Size size, Alignment alignment, \
+                                              SpareBitVector &&spareBits, \
+                                              bool isOptional) \
+      : SingleScalarTypeInfo(type, size, std::move(spareBits), \
+                             alignment, IsNotPOD, IsFixedSize), \
+        ValueTypeAndIsOptional(valueType, isOptional) {} \
+    enum { IsScalarPOD = false }; \
+    llvm::Type *getScalarType() const { \
+      return ValueTypeAndIsOptional.getPointer(); \
+    } \
+    Address projectScalar(IRGenFunction &IGF, Address addr) const { \
+      return IGF.Builder.CreateBitCast(addr, getScalarType()->getPointerTo()); \
+    } \
+    void emitScalarRetain(IRGenFunction &IGF, llvm::Value *value, \
+                          Atomicity atomicity) const { \
+      IGF.emitNative##Name##Retain(value, atomicity); \
+    } \
+    void emitScalarRelease(IRGenFunction &IGF, llvm::Value *value, \
+                           Atomicity atomicity) const { \
+      IGF.emitNative##Name##Release(value, atomicity); \
+    } \
+    void emitScalarFixLifetime(IRGenFunction &IGF, llvm::Value *value) const { \
+      IGF.emitFixLifetime(value); \
+    } \
+    unsigned getFixedExtraInhabitantCount(IRGenModule &IGM) const override { \
+      return IGM.get##Name##ExtraInhabitantCount(ReferenceCounting::Native) \
+          - ValueTypeAndIsOptional.getInt(); \
```

Comment:
I think this might be the bug — weak references have no extra inhabitants and always logically optional, so this subtraction. wraps around.

Reply:
If I understand you correctly, these comments imply that `getWeakExtraInhabitantCount()` is expected to return zero. Correct? If so, then the meta-programming of `getWeakExtraInhabitantCount()` is probably wrong then. Correct?

Reply:
PS – I was trying to generically allow for Unowned/Unmanaged to be optionally `Optional<T>`.

Reply:
Yeah, I don't think it's a good idea to meta-program `IGM::get##Name##ExtraInhabitantCount` unless you have a way to only cover the storages that are physically just normal references (like unmanaged, native-unowned, and presumably your new kinds).

Also, are you sure about `ReferenceCounting::Native` here?

Reply:
Okay. I'll look into that. Just to be clear, `getWeakExtraInhabitantCount()` is expected to return zero, despite weak being optional, right? What is native `getUnownedExtraInhabitantCount()` expected to return?

BTW – Unmanaged isn't guaranteed to be a normal reference. The clang importer creates unmanaged references that are also optional.

Reply:
Sorry, by "a normal reference" I meant "just a reference to the object, not decorated with additional stuff to implement its special storage semantics".  Optionality is separate.

Native unowned references have the extra-inhabitant count of the underlying reference, i.e. either `getHeapObjectExtraInhabitantCount` (if non-optional) or that - 1 (if optional).

Reply:
Hi @rjmccall – I think I figured out the problem with my attempt at consolidating the extra inhabitant logic between weak and unowned. Before this pull request, WeakTypeInfo implicitly inherited its extra inhabit logic from FixedTypeInfo, but Unowned explicitly overrode the extra inhabitant logic.

Therefore, before this pull request, we had the following scenario:

Weak+x86 (objc or native) == 0 extra inhabitants
Weak+x86_64 (objc or native) == INT_MAX extra inhabitants
Unowned+objc (x86 or x86_64) == 1 extra inhabitant
Unowned+native (x86 or x86_64) == HeapObject extra inhabitants

What if anything is the rational for the above differences?

For the record, this pull request consolidated the logic to be more like Unowned, therefore:

ObjC == 0 or 1 extra inhabitants depending on optionality
Native == HeapObject extra inhabitants

That being said, what should the default meta-programmed behavior be? Should Weak or Unowned override the default meta-programmed behavior and if so, then how?

Reply:
I would suggest that you avoid metaprogramming this, except maybe on the otherwise-orthogonal axis of "physically just an ordinary object pointer" like I mentioned before.  The ABI for what sort of extra inhabitants a storage kind is very much up to what the runtime wants to guarantee.

ObjC-compatible `unowned` should have one extra inhabitant when non-optional, none when optional.  `weak` should have zero.  `unowned(unsafe)` and non-ObjC `unowned` should have the extra inhabitants of their referent type (which might be optional).

Reply:
What does "this" refer to? Assuming that "loadable" implies "native"? Something else? In any case, I'll update the pull request shortly with the second paragraph feedback.

Reply:
Okay, I think I get it, and I think I can "de-meta-program" and let runtimes explicitly control the extra inhabitant policies for reference storage types. That being said, I'm still curious about why weak "should have zero" regardless of the runtime.